Flaky-test: ManagedCursorMetricsTest.testManagedCursorMetrics

Flaky-test: org.apache.pulsar.broker.stats.ManagedCursorMetricsTest.testManagedCursorMetrics
Number of failures: 10

org.apache.pulsar.broker.stats.ManagedCursorMetricsTest is flaky. The testManagedCursorMetrics test method fails sporadically.

```
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertTrue(Assert.java:45)
	at org.testng.Assert.assertTrue(Assert.java:55)
	at org.apache.pulsar.broker.stats.ManagedCursorMetricsTest.testManagedCursorMetrics(ManagedCursorMetricsTest.java:70)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.
[example failure 2021-03-08T17:30:26.3999781Z](https://github.com/apache/pulsar/runs/2056548871?check_suite_focus=true#step:8:190)
[example failure 2021-03-08T11:28:16.6508831Z](https://github.com/apache/pulsar/runs/2055117809?check_suite_focus=true#step:8:1036)
[example failure 2021-03-08T11:14:45.8989014Z](https://github.com/apache/pulsar/runs/2055117809?check_suite_focus=true#step:8:638)
[example failure 2021-03-05T21:58:01.2183193Z](https://github.com/apache/pulsar/runs/2043229259?check_suite_focus=true#step:8:190)
[example failure 2021-03-05T09:34:45.4232866Z](https://github.com/apache/pulsar/runs/2037997236?check_suite_focus=true#step:8:190)
[example failure 2021-03-05T09:22:25.4442285Z](https://github.com/apache/pulsar/runs/2037974738?check_suite_focus=true#step:8:1014)
[example failure 2021-03-05T09:10:53.5286506Z](https://github.com/apache/pulsar/runs/2037974738?check_suite_focus=true#step:8:614)
[example failure 2021-03-05T08:59:35.6870954Z](https://github.com/apache/pulsar/runs/2037974738?check_suite_focus=true#step:8:198)
[example failure 2021-03-04T06:23:51.0098464Z](https://github.com/apache/pulsar/runs/2028449927?check_suite_focus=true#step:8:688)
[example failure 2021-03-04T05:17:18.0634430Z](https://github.com/apache/pulsar/runs/2028174394?check_suite_focus=true#step:8:286)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertTrue(Assert.java:45)
	at org.testng.Assert.assertTrue(Assert.java:55)
	at org.apache.pulsar.broker.stats.ManagedCursorMetricsTest.testManagedCursorMetrics(ManagedCursorMetricsTest.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

recent failure: https://github.com/apache/pulsar/runs/5008959840?check_suite_focus=true#step:9:645
```
Error:  testManagedCursorMetrics(org.apache.pulsar.broker.stats.ManagedCursorMetricsTest)  Time elapsed: 0.238 s  <<< FAILURE!
java.lang.AssertionError: did not expect [0] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failEquals(Assert.java:1041)
	at org.testng.Assert.assertNotEqualsImpl(Assert.java:147)
	at org.testng.Assert.assertNotEquals(Assert.java:1531)
	at org.testng.Assert.assertNotEquals(Assert.java:1535)
	at org.apache.pulsar.broker.stats.ManagedCursorMetricsTest.testManagedCursorMetrics(ManagedCursorMetricsTest.java:89)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/ManagedCursorMetricsTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/ManagedCursorMetricsTest.java
@@ -18,14 +18,20 @@
  */
 package org.apache.pulsar.broker.stats;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import lombok.Cleanup;
-import org.apache.bookkeeper.client.PulsarMockLedgerHandle;
+import org.apache.bookkeeper.client.LedgerHandle;
+import org.apache.bookkeeper.mledger.ManagedCursorMXBean;
+import org.apache.bookkeeper.mledger.impl.ManagedCursorImpl;
 import org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest;
+import org.apache.pulsar.broker.service.persistent.PersistentSubscription;
 import org.apache.pulsar.broker.stats.metrics.ManagedCursorMetrics;
 import org.apache.pulsar.client.api.ClientBuilder;
 import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.client.api.PulsarClient;
 import org.apache.pulsar.client.api.PulsarClientException;
@@ -34,15 +40,16 @@ import org.apache.pulsar.client.impl.ConsumerImpl;
 import org.apache.pulsar.client.impl.PulsarTestClient;
 import org.apache.pulsar.common.stats.Metrics;
 import org.awaitility.Awaitility;
+import org.powermock.reflect.Whitebox;
 import org.testng.Assert;
-import org.testng.annotations.AfterMethod;
-import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
 @Test(groups = "broker")
 public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
 
-    @BeforeMethod(alwaysRun = true)
+    @BeforeClass
     @Override
     protected void setup() throws Exception {
         conf.setTopicLevelPoliciesEnabled(false);
@@ -50,7 +57,7 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
         super.internalSetup();
     }
 
-    @AfterMethod(alwaysRun = true)
+    @AfterClass
     @Override
     protected void cleanup() throws Exception {
         super.internalCleanup();
@@ -61,21 +68,28 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
         return PulsarTestClient.create(clientBuilder);
     }
 
+    /***
+     * This method has overridden these case:
+     *    brk_ml_cursor_persistLedgerSucceed
+     *    brk_ml_cursor_persistLedgerErrors
+     *    brk_ml_cursor_persistZookeeperSucceed
+     *    brk_ml_cursor_nonContiguousDeletedMessagesRange
+     * But not overridden "brk_ml_cursor_nonContiguousDeletedMessagesRange".
+     */
     @Test
     public void testManagedCursorMetrics() throws Exception {
         final String subName = "my-sub";
         final String topicName = "persistent://my-namespace/use/my-ns/my-topic1";
-        final int messageSize = 10;
-
+        /** Before create cursor. Verify metrics will not be generated. **/
+        // Create ManagedCursorMetrics and verify empty.
         ManagedCursorMetrics metrics = new ManagedCursorMetrics(pulsar);
-
         List<Metrics> metricsList = metrics.generate();
         Assert.assertTrue(metricsList.isEmpty());
-
-        metricsList = metrics.generate();
-        Assert.assertTrue(metricsList.isEmpty());
-
-        PulsarTestClient pulsarClient = (PulsarTestClient) this.pulsarClient;
+        /**
+         * Trigger the cursor ledger creation.
+         * After create cursor. Verify all metrics is zero.
+         */
+        // Trigger cursor creation.
         @Cleanup
         ConsumerImpl<byte[]> consumer = (ConsumerImpl<byte[]>) this.pulsarClient.newConsumer()
                 .topic(topicName)
@@ -84,28 +98,84 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
                 .subscriptionName(subName)
                 .isAckReceiptEnabled(true)
                 .subscribe();
-
-
         @Cleanup
         Producer<byte[]> producer = this.pulsarClient.newProducer()
                 .topic(topicName)
+                .enableBatching(false)
                 .create();
-
-        producer.send("trigger-cursor-ledger-creation".getBytes());
-        // Trigger the cursor ledger creation
-        consumer.acknowledge(consumer.receive().getMessageId());
-
-        for(PulsarMockLedgerHandle ledgerHandle : mockBookKeeper.getLedgerMap().values()) {
-            ledgerHandle.close();
-        }
-
-        for (int i = 0; i < messageSize; i++) {
+        final PersistentSubscription persistentSubscription =
+                (PersistentSubscription) pulsar.getBrokerService()
+                        .getTopic(topicName, false).get().get().getSubscription(subName);
+        final ManagedCursorImpl managedCursor = (ManagedCursorImpl) persistentSubscription.getCursor();
+        ManagedCursorMXBean managedCursorMXBean = managedCursor.getStats();
+        // Assert.
+        metricsList = metrics.generate();
+        Assert.assertFalse(metricsList.isEmpty());
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerErrors"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperErrors"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_nonContiguousDeletedMessagesRange"), 0L);
+        /**
+         * 1. Send many messages, and only ack half. After the cursor data is written to BK,
+         *    verify "brk_ml_cursor_persistLedgerSucceed" and "brk_ml_cursor_nonContiguousDeletedMessagesRange".
+         * 2. Ack another half, verify "brk_ml_cursor_nonContiguousDeletedMessagesRange" is zero.
+         */
+        // Send many message and ack half.
+        List<MessageId> keepsMessageIdList = new ArrayList<>();
+        for (int i = 0; i < 30; i++) {
             String message = "my-message-" + i;
             producer.send(message.getBytes());
-            consumer.acknowledge(consumer.receive().getMessageId());
+            if (i < 10 || i > 20) {
+                consumer.acknowledge(consumer.receive().getMessageId());
+            } else {
+                keepsMessageIdList.add(consumer.receive().getMessageId());
+            }
         }
-
-        Awaitility.await().until(() -> pulsarClient.getConnection(topicName).get().getPendingRequests().size() == 0);
+        // Wait persistent.
+        Awaitility.await().atMost(2, TimeUnit.SECONDS)
+                .until(() -> managedCursorMXBean.getPersistLedgerSucceed() > 0);
+        // Assert.
+        metricsList = metrics.generate();
+        Assert.assertFalse(metricsList.isEmpty());
+        Assert.assertNotEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerErrors"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperErrors"), 0L);
+        Assert.assertNotEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_nonContiguousDeletedMessagesRange"),
+                0L);
+        // Ack another half.
+        for (MessageId messageId : keepsMessageIdList){
+            consumer.acknowledge(messageId);
+        }
+        // Wait persistent.
+        Awaitility.await().atMost(2, TimeUnit.SECONDS)
+                .until(() -> managedCursor.getTotalNonContiguousDeletedMessagesRange() == 0);
+        // Assert.
+        metricsList = metrics.generate();
+        Assert.assertFalse(metricsList.isEmpty());
+        Assert.assertNotEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerErrors"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperSucceed"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperErrors"), 0L);
+        Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_nonContiguousDeletedMessagesRange"), 0L);
+        /**
+         * Make BK error, and send many message, then wait cursor persistent finish.
+         * After the cursor data is written to ZK, verify "brk_ml_cursor_persistLedgerErrors" and
+         * "brk_ml_cursor_persistZookeeperSucceed".
+         */
+        // Send amd ack messages, at the same time makes BK error.
+        Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> {
+            String message = UUID.randomUUID().toString();
+            producer.send(message.getBytes());
+            consumer.acknowledge(consumer.receive().getMessageId());
+            // Make BK error.
+            LedgerHandle ledgerHandle = Whitebox.getInternalState(managedCursor, "cursorLedger");
+            ledgerHandle.close();
+            return managedCursorMXBean.getPersistLedgerErrors() > 0
+                    && managedCursorMXBean.getPersistZookeeperSucceed() > 0;
+        });
+        // Assert.
         metricsList = metrics.generate();
         Assert.assertFalse(metricsList.isEmpty());
         Assert.assertNotEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistLedgerSucceed"), 0L);
@@ -113,6 +183,18 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
         Assert.assertNotEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperSucceed"), 0L);
         Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_persistZookeeperErrors"), 0L);
         Assert.assertEquals(metricsList.get(0).getMetrics().get("brk_ml_cursor_nonContiguousDeletedMessagesRange"), 0L);
+        /**
+         * TODO verify "brk_ml_cursor_persistZookeeperErrors".
+         * This is not easy to implement, we can use {@link #mockZooKeeper} to fail ZK, but we cannot identify whether
+         * the request is triggered by the "create new ledger then write ZK" or the "persistent cursor then write ZK".
+         * The cursor path is "/managed-ledgers/my-namespace/use/my-ns/persistent/my-topic1/my-sub". Maybe we can
+         * mock/spy ManagedCursorImpl to overridden this case in another PR.
+         */
+        mockZooKeeper.unsetAlwaysFail();
+        producer.close();
+        consumer.close();
+        managedCursor.close();
+        admin.topics().delete(topicName, true);
     }
 
     @Test
@@ -150,10 +232,6 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
                 .topic(topicName)
                 .create();
 
-        for (PulsarMockLedgerHandle ledgerHandle : mockBookKeeper.getLedgerMap().values()) {
-            ledgerHandle.close();
-        }
-
         for (int i = 0; i < messageSize; i++) {
             String message = "my-message-" + i;
             producer.send(message.getBytes());
@@ -172,5 +250,11 @@ public class ManagedCursorMetricsTest extends MockedPulsarServiceBaseTest {
         Assert.assertNotEquals(metricsList.get(1).getMetrics().get("brk_ml_cursor_writeLedgerSize"), 0L);
         Assert.assertNotEquals(metricsList.get(1).getMetrics().get("brk_ml_cursor_writeLedgerLogicalSize"), 0L);
         Assert.assertEquals(metricsList.get(1).getMetrics().get("brk_ml_cursor_readLedgerSize"), 0L);
+
+        // cleanup.
+        consumer.close();
+        consumer2.close();
+        producer.close();
+        admin.topics().delete(topicName, true);
     }
 }

